File: algorithms/easy/divide_array_into_equal_pairs.py
from typing import List
from collections import Counter

class Solution:
    def divideArray(self, nums: List[int]) -> bool:
        # TC = O(N), SC = O(N)
        unmatched = set()
        for n in nums:
            if n in unmatched:
                unmatched.remove(n)
            else:
                unmatched.add(n)
        # There should be no unmatched integer in the end
        return len(unmatched) == 0

    def divideArray_1(self, nums: List[int]) -> bool:
        # TC = O(N), SC = O(N)
        cntr = Counter(nums)
        for v in cntr.values():
            if v % 2 == 1:
                return False
        return True

    def divideArray_2(self, nums: List[int]) -> bool:
        # TC = O(NLogN), SC = O(1)
        nums.sort()
        for i in range(0, len(nums), 2):
            if nums[i] != nums[i+1]:
                return False
        return True

# Main section
for nums in [
               [3,2,3,2,2,2],
               [1,2,3,4],
            ]:
    print(f'nums = {nums}')
    sol = Solution()
    r = sol.divideArray(nums)
    r1 = sol.divideArray_1(nums)
    r2 = sol.divideArray_2(nums)
    print(f'r  = {r}')
    print(f'r1 = {r1}')
    print(f'r2 = {r2}')
    print('================')

File: algorithms/easy/divide_array_into_equal_pairs.go
package main

import (
          "fmt"
          "sort"
       )

func divideArray(nums []int) bool {
    hsh := make(map[int]int)
    for _, n := range nums {
        _, ok := hsh[n]
        if ok {
            delete(hsh, n)
        } else {
            hsh[n]++
        }
    }
    return len(hsh) == 0
}

func divideArray_1(nums []int) bool {
    hsh := make(map[int]int)
    for _, v := range nums {
        hsh[v]++
    }
    for _, v := range hsh {
        if v % 2 == 1 {
            return false
        }
    }
    return true
}

func divideArray_2(nums []int) bool {
    N := len(nums)
    sort.Ints(nums)
    for i := 0; i < N; i += 2 {
        if nums[i] != nums[i+1] {
            return false
        }
    }
    return true
}

func main() {
    data := [][]int {
                []int{3,2,3,2,2,2},
                []int{1,2,3,4},
            }
    for _, nums := range data {
        fmt.Printf("nums = %v\n", nums)
        r := divideArray(nums)
        r1 := divideArray_1(nums)
        r2 := divideArray_2(nums)
        fmt.Printf("r  = %v\n", r)
        fmt.Printf("r1 = %v\n", r1)
        fmt.Printf("r2 = %v\n", r2)
        fmt.Println("======================")
    }
}

